### PR TITLE
Cover shipping method management

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13                   |
| Bug fix?        | no                                                       |
| New feature?    | yes                                                       |
| BC breaks?      | no                                                      |
| License         | MIT                                                          |

### CODE REVIEW

File: config/packages/_sylius.yaml [hunk 1]
```diff
@@ -22,6 +22,28 @@ sylius_payum:
             - 'sylius'
             - 'stripe_checkout'
 
+sylius_shipping:
+    shipping_method_rules_validation_groups:
+        total_weight_greater_than_or_equal:
+            - 'sylius'
+            - 'total_weight'
```
Comment (Wojdylak) [inline config/packages/_sylius.yaml]: `sylius_` prefix WDYT?

File: features/shipping/managing_shipping_methods/shipping_method_validation.feature [hunk 2]
```diff
@@ -46,10 +46,18 @@ Feature: Shipping method validation
...
+    @ui @no-api
     Scenario: Trying to remove zone from existing shipping method
         Given the store allows shipping with "UPS Ground"
         When I want to modify this shipping method
         And I remove its zone
         And I try to save my changes
         Then I should be notified that zone has to be selected
+
+    @no-ui @api
+    Scenario: Trying to remove zone from existing shipping method
```
Comment (Wojdylak) [inline features/shipping/managing_shipping_methods/shipping_method_validation.feature]: Why don't you connect these scenarios, changing only the name of the last step?

File: src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml [hunk 3]
```diff
@@ -16,6 +16,7 @@ parameters:
         - "%sylius.security.new_api_route%/admin/promotion-actions/{id}"
         - "%sylius.security.new_api_route%/admin/promotion-rules/{id}"
         - "%sylius.security.new_api_route%/admin/shop-users/{id}"
+        - "%sylius.security.new_api_route%/admin/shipping-method-rules/{id}"
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml]:
```suggestion
        - "%sylius.security.new_api_route%/admin/shipping-method-rules/{id}"
        - "%sylius.security.new_api_route%/admin/shop-users/{id}"
```
It would be nice to keep alphabetical order

File: features/shipping/managing_shipping_methods/shipping_method_code_validation.feature [hunk 4]
```diff
@@ -28,8 +28,8 @@ Feature: Shipping method code validation
         Then I should be notified that code needs to contain only specific symbols
         And shipping method with name "FedEx Carrier" should not be added
 
-    @ui @javascript
-    Scenario: Trying to add a new shipping method with
+    @ui @javascript @api
+    Scenario: Trying to add a new shipping method with valid code
```
Comment (NoResponseMate) [inline features/shipping/managing_shipping_methods/shipping_method_code_validation.feature]: This whole scenario can be removed, there's no validation present

File: tests/Api/Admin/ShippingMethodsTest.php [hunk 5]
```diff
@@ -97,7 +97,7 @@ public function it_archives_a_shipping_method(): void
     }
 
     /** @test */
-    public function it_restores_a_shipping_method(): void
```
Comment (NoResponseMate) [inline tests/Api/Admin/ShippingMethodsTest.php]: These were valid, why did you overwrite them?

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/ShippingMethodRule.xml [hunk 6]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" ?>
+
+<!--
+ This file is part of the Sylius package.
+ (c) Sylius Sp. z o.o.
+ For the full copyright and license information, please view the LICENSE
+ file that was distributed with this source code.
+-->
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/Resources/config/serialization/ShippingMethodRule.xml]: Bit tight ain't it

File: config/packages/_sylius.yaml [hunk 7]
```diff
@@ -22,6 +22,28 @@ sylius_payum:
             - 'sylius'
             - 'stripe_checkout'
 
+sylius_shipping:
+    shipping_method_rules_validation_groups:
+        total_weight_greater_than_or_equal:
```
Comment (NoResponseMate) [inline config/packages/_sylius.yaml]:
```suggestion
sylius_shipping:
    shipping_method_rule:
        validation_groups:    
            total_weight_greater_than_or_equal:
```
vs
```suggestion
sylius_shipping:
    shipping_method_rules_validation_groups:
        total_weight_greater_than_or_equal:
```
We need to decide on one, gateway config uses the latter, while the WIP promotions use the former.

File: config/packages/_sylius.yaml (hunk 1 above)
Comment (Wojdylak) [reply config/packages/_sylius.yaml]: or https://github.com/Sylius/Sylius/pull/15628#discussion_r1425585555

File: src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php [hunk 8]
```diff
@@ -578,6 +719,72 @@ public function iShouldBeViewingNonArchivalShippingMethods(): void
...
+        Assert::contains(
+            $this->responseChecker->getError($this->client->getLastResponse()),
+            sprintf('configuration[%s][amount]: This value should not be blank.', $channel->getCode()),
+        );
+    }
+
+    /**
+     * @Then I should be notified that code needs to contain only specific symbols
+     */
+    public function iShouldBeNotifiedThatCodeShouldContain(): void
```
Comment (Wojdylak) [inline src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php]:
```suggestion
    public function iShouldBeNotifiedThatCodeNeedsToContainOnlySpecificSymbols(): void
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php [hunk 9]
```diff
@@ -578,6 +719,72 @@ public function iShouldBeViewingNonArchivalShippingMethods(): void
...
+        Assert::contains(
+            $this->responseChecker->getError($this->client->getLastResponse()),
+            'rules[0].configuration[weight]: This value should be of type numeric.',
+        );
+    }
+
+    /**
+     * @Then I should be notified that the amount rule has an invalid configuration in :channel channel
+     */
+    public function iShouldBeNotifiedThatTheAmountRuleHasAnInvalidConfiguration(ChannelInterface $channel): void
```
Comment (Wojdylak) [inline src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php]:
```suggestion
    public function iShouldBeNotifiedThatTheAmountRuleHasAnInvalidConfigurationInChannel(ChannelInterface $channel): void
```

File: features/shipping/managing_shipping_methods/shipping_method_validation.feature (hunk 2 above)
Comment (TheMilek) [reply features/shipping/managing_shipping_methods/shipping_method_validation.feature]: I mean the problem is the step `Then I should be notified that zone has to be selected` its really specific for the ui, I mean I dont really like adjusting the scenarios by force to cover both contexts, thats why i separated them, as the info on the API site is saying about that not being blank and its impossible to select blank choice from the select dropdown

File: tests/Api/Admin/ShippingMethodsTest.php (hunk 5 above)
Comment (TheMilek) [reply tests/Api/Admin/ShippingMethodsTest.php]: I dont know how it happened must admit 🤷🏽‍♂️

File: features/shipping/managing_shipping_methods/shipping_method_validation.feature (hunk 2 above)
Comment (NoResponseMate) [reply features/shipping/managing_shipping_methods/shipping_method_validation.feature]: You could just change the step to something like
> I should be notified that the zone is required

It's nebulous enough to fit both cases

File: UPGRADE-1.13.md [hunk 10]
```diff
@@ -18,6 +18,28 @@ sylius_payum:
...
+        your_rule:
+            - 'sylius'
+            - 'your_custom_validation_group'
+    shipping_method_calculators_validation_groups:
+        flat_rate:
+            - 'sylius'
+            - 'rate'
+        your_calculator:
+            - 'sylius'
+            - 'your_custom_validation_group'
```
Comment (NoResponseMate) [inline UPGRADE-1.13.md]: Remember to update these when the actual groups change

File: src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php [hunk 11]
```diff
@@ -63,6 +64,102 @@ public function iSortShippingMethodsByName(string $sortType = 'ascending'): void
         ]);
     }
 
+    /**
+     * @When I add the "Total weight less than or equal" rule configured with :weight
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php]: I feel like `Total weight less/greater than or equal` could parametrized and stringified into the type, might reduce the number of functions a bit.

File: src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php [hunk 12]
```diff
@@ -123,6 +220,28 @@ public function iTryToCreateANewShippingMethodWithValidData(): void
...
+     */
+    public function iDoNotSpecifyAmountForCalculator(string $calculatorName): void
+    {
+        $channelCode = $this->sharedStorage->get('channel')->getCode();
+        match ($calculatorName) {
+            'Flat rate per shipment' => $this->client->addRequestData('calculator', DefaultCalculators::FLAT_RATE),
+            'Flat rate per unit' => $this->client->addRequestData('calculator', DefaultCalculators::PER_UNIT_RATE),
+        };
+
+        $this->client->addRequestData('configuration', [$channelCode => ['amount' => null]]);
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php]:
```suggestion
        match ($calculatorName) {
            'Flat rate per shipment' => $this->client->addRequestData('calculator', DefaultCalculators::FLAT_RATE),
            'Flat rate per unit' => $this->client->addRequestData('calculator', DefaultCalculators::PER_UNIT_RATE),
        };

        $channelCode = $this->sharedStorage->get('channel')->getCode();
        $this->client->addRequestData('configuration', [$channelCode => ['amount' => null]]);
```
You could also throw on the default match since no known calculator has been passed.

File: src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php [hunk 13]
```diff
@@ -578,6 +719,72 @@ public function iShouldBeViewingNonArchivalShippingMethods(): void
...
+    }
+
+    /**
+     * @Then I should be notified that the weight rule has an invalid configuration
+     */
+    public function iShouldBeNotifiedThatTheWeightRuleHasAnInvalidConfiguration(): void
+    {
+        Assert::contains(
+            $this->responseChecker->getError($this->client->getLastResponse()),
+            'rules[0].configuration[weight]: This value should be of type numeric.',
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php]:
```suggestion
            'configuration[weight]: This value should be of type numeric.',
```
As is it cannot be reused in scenarios with multiple rules

File: src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php [hunk 14]
```diff
@@ -578,6 +719,72 @@ public function iShouldBeViewingNonArchivalShippingMethods(): void
...
+    }
+
+    /**
+     * @Then I should be notified that the amount rule has an invalid configuration in :channel channel
+     */
+    public function iShouldBeNotifiedThatTheAmountRuleHasAnInvalidConfigurationInChannel(ChannelInterface $channel): void
+    {
+        Assert::contains(
+            $this->responseChecker->getError($this->client->getLastResponse()),
+            sprintf('rules[0].configuration[%s][amount]: This value should be of type numeric.', $channel->getCode()),
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php]:
```suggestion
            sprintf('configuration[%s][amount]: This value should be of type numeric.', $channel->getCode()),
```

File: src/Sylius/Behat/Page/Admin/ShippingMethod/CreatePage.php [hunk 15]
```diff
@@ -126,13 +120,33 @@ protected function getDefinedElements(): array
             'channel' => '#sylius_shipping_method_channels .ui.checkbox:contains("%channel%")',
             'calculator' => '#sylius_shipping_method_calculator',
             'calculator_configuration' => '.ui.segment.configuration',
+            'weight' => '#sylius_shipping_method_rules_0_configuration_weight',
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Page/Admin/ShippingMethod/CreatePage.php]: Should be refactored to not rely on the index

File: features/shipping/managing_shipping_methods/editing_shipping_method.feature [hunk 16]
```diff
@@ -10,14 +10,6 @@ Feature: Editing shipping method
...
         And I am logged in as an administrator
 
-    @todo
-    Scenario: Trying to change shipping method code
-        When I want to modify a shipping method "UPS Carrier"
-        And I change its code to "UPS"
-        And I save my changes
-        Then I should be notified that code cannot be changed
-        And shipping method "UPS Carrier" should still have code "UPS_CARRIER"
-
```
Comment (jakubtobiasz) [inline features/shipping/managing_shipping_methods/editing_shipping_method.feature:20]: I don't see this scenario anywhere now, why did you remove it?

File: src/Sylius/Bundle/ShippingBundle/Resources/config/services/validator.xml [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+ file that was distributed with this source code.
+
+-->
+
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd"
+>
+    <services>
+        <defaults public="true" />
```
Comment (jakubtobiasz) [inline src/Sylius/Bundle/ShippingBundle/Resources/config/services/validator.xml]: Maybe we should start finally **not** making all services public by default. It makes no sense. I **know** that **we used to do it that way**, but as I can understand keeping it in already added services for new ones is pointless.

**Review by jakubtobiasz — DISMISSED** [review]

File: features/shipping/managing_shipping_methods/editing_shipping_method.feature (hunk 16 above)
Comment (NoResponseMate) [reply features/shipping/managing_shipping_methods/editing_shipping_method.feature:20]: It's basically the one right below

File: config/packages/_sylius.yaml [hunk 18]
```diff
@@ -22,6 +22,30 @@ sylius_payum:
             - 'sylius'
             - 'stripe_checkout'
 
+sylius_shipping:
```
Comment (diimpp) [inline config/packages/_sylius.yaml]: Would be cool to sync definitions with recently merged promotion validation groups for actions/rules.

image
https://github.com/Sylius/Sylius/pull/15628/files#diff-2ba8309ba302fddcdc496c9644ca140fd71904b9fd60edabafd2af0822a1942f

File: config/packages/_sylius.yaml [hunk 19]
```diff
@@ -22,6 +22,30 @@ sylius_payum:
...
+    shipping_method_rule:
+        validation_groups:
+            total_weight_greater_than_or_equal:
+                - 'sylius'
+                - 'sylius_total_weight'
```
Comment (diimpp) [inline config/packages/_sylius.yaml]:
```suggestion
                - 'sylius_shipping_method_rule_total_weight'
```

File: config/packages/_sylius.yaml [hunk 20]
```diff
@@ -22,6 +22,30 @@ sylius_payum:
...
+                - 'sylius'
+                - 'sylius_order_total'
+            order_total_less_than_or_equal:
+                - 'sylius'
+                - 'sylius_order_total'
+    shipping_method_calculator:
+        validation_groups:
+            flat_rate:
+                - 'sylius'
+                - 'sylius_rate'
```
Comment (diimpp) [inline config/packages/_sylius.yaml]:
```suggestion
                - 'sylius_shipping_method_calculator_rate'
```

File: UPGRADE-1.13.md [hunk 21]
```diff
@@ -18,6 +18,28 @@ sylius_payum:
             - 'your_custom_validation_group'
 ```
 
+1. There have been a naw parameters added to specify the validation groups for given shipping method rule and calculator.
+   If you have any custom validation groups for your calculator or rules, you need to add them to your `config/packages/_sylius.yaml` file.
+   Also, if you have your own shipping method rule or calculator and want to add your validation groups you can add another key to the `shipping_method_rules_validation_groups` or `shipping_method_calculators_validation_groups` parameter.
+
+```yaml
+sylius_shipping:
+    shipping_method_rules_validation_groups:
```
Comment (diimpp) [inline UPGRADE-1.13.md]:
```suggestion
sylius_shipping:
    shipping_method_rule:
        validation_groups:
```

File: UPGRADE-1.13.md [hunk 22]
```diff
@@ -18,6 +18,28 @@ sylius_payum:
...
+        order_total_greater_than_or_equal:
+            - 'sylius'
+            - 'order_total'
+        your_rule:
+            - 'sylius'
+            - 'your_custom_validation_group'
+    shipping_method_calculators_validation_groups:
```
Comment (diimpp) [inline UPGRADE-1.13.md]:
```suggestion
    shipping_method_calculator:
        validation_groups:
```

File: features/shipping/managing_shipping_methods/adding_shipping_method_with_rule.feature [hunk 23]
```diff
@@ -65,3 +65,31 @@ Feature: Adding a new shipping method with rule
         And I add it
         Then I should be notified that it has been successfully created
         And the shipping method "FedEx Carrier" should appear in the registry
+
+    @ui @javascript @api
```
Comment (diimpp) [inline features/shipping/managing_shipping_methods/adding_shipping_method_with_rule.feature]: Please specify rules types in both scenario headers, like it's done for success cases above.

Comment (diimpp) [reply features/shipping/managing_shipping_methods/adding_shipping_method_with_rule.feature]: Feels like those scenarios should be placed at `features/shipping/managing_shipping_methods/shipping_method_validation.feature`

**Review by jakubtobiasz — DISMISSED** [review]

File: features/shipping/managing_shipping_methods/adding_shipping_method_with_rule.feature [hunk 24]
```diff
@@ -61,7 +61,7 @@ Feature: Adding a new shipping method with rule
         And I define it for the zone named "United States"
         And I choose "Flat rate per shipment" calculator
         And I specify its amount as 50 for "Web-US" channel
-        And I add the "Items total less than or equal" rule configured with $200 for "Web-US" channel
+        And I add the Items total less than or equal rule configured with $200 for "Web-US" channel
```
Comment (NoResponseMate) [inline features/shipping/managing_shipping_methods/adding_shipping_method_with_rule.feature]:
```suggestion
        And I add the "Items total less than or equal" rule configured with $200 for "Web-US" channel
```
It's easier to read when it's in quotes

File: src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php [hunk 25]
```diff
@@ -63,6 +64,94 @@ public function iSortShippingMethodsByName(string $sortType = 'ascending'): void
...
+            'Total weight greater than or equal' => $this->client->addRequestData('rules', [
+                [
+                    'type' => 'total_weight_greater_than_or_equal',
+                    'configuration' => [
+                        'weight' => $weight,
+                    ],
+                ],
+            ]),
+            default => throw new \InvalidArgumentException('Unsupported shipping method rule'),
+        };
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php]:
```suggestion
        $type = StringInflector::nameToLowercaseCode($rule);
        $this->client->addRequestData('rules', [[
            'type' => $type,
            'configuration' => [
                'weight' => $weight,
            ],
        ]]);
```
and if you prefer to leave the match, refactor it to resolve the `type` by itself and add that into the request data. Similar with `::iAddTheRuleConfiguredWithForChannel`.<br>
Also, after refactoring you could extract the part which adds the data into a private method; the `$this->client->addRequestData('rules', [` is used a lot in here

File: src/Sylius/Behat/Context/Ui/Admin/ManagingShippingMethodsContext.php [hunk 26]
```diff
@@ -504,13 +525,23 @@ public function iAddTheItemsTotalGreaterThanOrEqualRuleConfiguredWith($value, Ch
...
     }
 
+    /**
+     * @When /^I add the "Items total less than or equal" rule configured with invalid data for ("[^"]+" channel)$/
+     */
+    public function iAddTheItemsTotalLessThanOrEqualRuleConfiguredWithInvalidData(ChannelInterface $channel): void
+    {
+        $this->createPage->addRule('Items total less than or equal');
+        $this->createPage->fillRuleOptionForChannel($channel->getCode(), 'Amount', 'Invalid data');
+    }
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingShippingMethodsContext.php]: That's basically the same thing; pretty sure you can just merge them

File: src/Sylius/Bundle/ShippingBundle/Resources/translations/validators.en.yml [hunk 27]
```diff
@@ -16,6 +16,7 @@ sylius:
         calculator:
             min: 'Shipping charge cannot be lower than 0.'
             not_blank: 'Please select shipping method calculator.'
+            invalid_calculator: 'Invalid calculator. Available calculators are {{ available_calculators }}.'
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ShippingBundle/Resources/translations/validators.en.yml]:
```suggestion
            invalid: 'Invalid calculator. Available calculators are {{ available_calculators }}.'
```

**Review by Wojdylak — DISMISSED** [review]

File: src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php [hunk 28]
```diff
@@ -63,6 +64,94 @@ public function iSortShippingMethodsByName(string $sortType = 'ascending'): void
...
+
+    /**
+     * @When /^I add the ([^"]+) rule configured with (?:€|£|\$)([^"]+) for ("[^"]+" channel)$/
+     */
+    public function iAddTheRuleConfiguredWithForChannel(string $rule, int $value, ChannelInterface $channel): void
+    {
+        match ($rule) {
+            'Items total less than or equal' => $this->client->addRequestData('rules', [
+                [
+                    'type' => 'order_total_less_than_or_equal',
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php:105]: Sidenote;
We really need to refactor this, the label has the context of what is actually taken into account, while the type is just wrong. The checker compares the items' total instead of the order's

File: src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php [hunk 29]
```diff
@@ -63,6 +65,82 @@ public function iSortShippingMethodsByName(string $sortType = 'ascending'): void
...
+            ],
+        ]);
+    }
+
+    /**
+     * @When /^I add the "([^"]+)" rule configured with (?:€|£|\$)([^"]+) for ("[^"]+" channel)$/
+     */
+    public function iAddTheRuleConfiguredWithForChannel(string $rule, int $value, ChannelInterface $channel): void
+    {
+        match ($rule) {
```
Comment (Rafikooo) [inline src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php:102]: The `match` could match and assign a local variable with the correct `type` and then the `addRequestData` function could be called below

File: src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php [hunk 30]
```diff
@@ -63,6 +65,82 @@ public function iSortShippingMethodsByName(string $sortType = 'ascending'): void
...
+    }
+
+    /**
+     * @When /^I add the "Items total less than or equal" rule configured with invalid data for ("[^"]+" channel)$/
+     */
+    public function iAddTheItemsTotalLessThanOrEqualRuleConfiguredWithInvalidData(ChannelInterface $channel): void
+    {
+        $this->client->addRequestData('rules', [
+            [
+                'type' => 'order_total_greater_than_or_equal',
```
Comment (Rafikooo) [inline src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php:134]:
```suggestion
                'type' => 'order_total_less_than_or_equal',
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php [hunk 31]
```diff
@@ -123,6 +201,29 @@ public function iTryToCreateANewShippingMethodWithValidData(): void
...
+    public function iDoNotSpecifyAmountForCalculator(string $calculatorName): void
+    {
+        match ($calculatorName) {
+            'Flat rate per shipment' => $this->client->addRequestData('calculator', DefaultCalculators::FLAT_RATE),
+            'Flat rate per unit' => $this->client->addRequestData('calculator', DefaultCalculators::PER_UNIT_RATE),
+            'default' => throw new \InvalidArgumentException('Unsupported calculator name'),
+        };
+
+        $channelCode = $this->sharedStorage->get('channel')->getCode();
+        $this->client->addRequestData('configuration', [$channelCode => ['amount' => null]]);
```
Comment (Rafikooo) [inline src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php:216]: What if we don't pass this key at all?

File: src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php [hunk 32]
```diff
@@ -527,6 +639,17 @@ public function iShouldBeNotifiedThatZoneHasToBeSelected(): void
...
     }
 
+    /**
+     * @Then I should be notified that the zone is required
+     */
+    public function iShouldBeNotifiedThatZoneHasToBeIriAndCannotBeNull(): void
+    {
+        Assert::contains(
+            $this->responseChecker->getError($this->client->getLastResponse()),
+            'Expected IRI or nested document for attribute "zone", "NULL" given.',
```
Comment (Rafikooo) [inline src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php:649]: Mixed statements

**Review by jakubtobiasz — APPROVED** [review]

**Comment by jakubtobiasz** [comment]
Thank you, @TheMilek!

File: src/Sylius/Behat/Context/Ui/Admin/ManagingShippingMethodsContext.php [hunk 33]
```diff
@@ -146,6 +146,16 @@ public function theShipmentMethodShouldAppearInTheRegistry(string $shipmentMetho
         Assert::true($this->indexPage->isSingleResourceOnPage(['name' => $shipmentMethodName]));
     }
 
+    /**
+     * @Then the shipping method :shipmentMethodName should not appear in the registry
+     */
+    public function theShipmentMethodShouldNotAppearInTheRegistry(string $shipmentMethodName): void
+    {
+        $this->iWantToBrowseShippingMethods();
```
Comment (Rafikooo) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingShippingMethodsContext.php:154]:
```suggestion
        $this->indexPage->open();
```
wdyt?

File: src/Sylius/Behat/Page/Admin/ShippingMethod/CreatePage.php [hunk 34]
```diff
@@ -126,13 +120,33 @@ protected function getDefinedElements(): array
...
             'zone' => '#sylius_shipping_method_zone',
             'rules' => '#sylius_shipping_method_rules',
         ]);
     }
 
+    private function getValidationMessageForElement(NodeElement $element): string
+    {
+        if (null === $element) {
+            throw new ElementNotFoundException($this->getSession(), 'Field element');
+        }
```
Comment (Rafikooo) [inline src/Sylius/Behat/Page/Admin/ShippingMethod/CreatePage.php:135]:
```suggestion
    private function getValidationMessageForElement(?NodeElement $element = null): string
    {
        if (null === $element) {
            throw new ElementNotFoundException($this->getSession(), 'Field element');
        }
```
As we made it nullable here `private function getChannelConfigurationOfLastRule(string $channelCode): ?NodeElement` as well

File: src/Sylius/Bundle/ShippingBundle/Resources/config/services/validator.xml [hunk 35]
```diff
@@ -0,0 +1,34 @@
...
+ For the full copyright and license information, please view the LICENSE
+ file that was distributed with this source code.
+
+-->
+
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd"
+>
+    <services>
```
Comment (Rafikooo) [inline src/Sylius/Bundle/ShippingBundle/Resources/config/services/validator.xml:18]: The 'public' attribute could be moved above or possibly removed if I'm not mistaken

**Review by Rafikooo — APPROVED** [review]
My comments were just some minor code style concerns that could be improved later
